core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningManager.java

```diff
@@ -0,0 +1,355 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.operator.exchange;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.log.Logger;
+import io.airlift.units.DataSize;
+import org.assertj.core.util.VisibleForTesting;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static java.lang.Double.isNaN;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Helps in finding skewness across writers and partitions when writing partitioned data using preferred partitioning.
+ * It then tries to distribute those skewed partitions across multiple writers.
+ * <p>
+ * Example:
+ * <p>
+ * Before: For three writers with skewed partitions
+ *      Writer 1 -> No partition assigned -> 0 bytes
+ *      Writer 2 -> No partition assigned -> 0 bytes
+ *      Writer 3 -> Partition 1 (100MB) + Partition 2 (100MB) + Partition 3 (100MB) ->  300 MB
+ * <p>
+ * After scaling:
+ *      Writer 1 -> Partition 1 (50MB) + Partition 3 (50MB) -> 100 MB
+ *      Writer 2 -> Partition 2 (50MB) -> 50 MB
+ *      Writer 3 -> Partition 1 (150MB) + Partition 2 (150MB) + Partition 3 (150MB) ->  450 MB
+ */
+public class ScaleWriterPartitioningManager
+{
+    private static final Logger log = Logger.get(ScaleWriterPartitioningManager.class);
+
+    private static final long MIN_SCALING_TRIGGER_BYTES = DataSize.of(100, MEGABYTE).toBytes();
+    private static final double SKEWNESS_THRESHOLD = 0.7;
+    private static final int MAX_PARTITIONS_TO_SCALE = 100;
```

Comment:
why to add this limit? What if skewed partitions change over time? Some javadoc woudl be great

Reply:
removed